### ISSUE TITLE
Support mfa without assume role.

### ISSUE THREAD

**Issue by glennpratt** [aws/aws-cli#1985]
Given the following profile:

``` ini
[profile test]
region = us-east-1
mfa_serial = arn:aws:iam::012345678901:mfa/glennpratt
```

aws-cli never asks for an MFA token and I don't receive IAM permissions granted to MFA users. Here's the policy:

``` json
{  
    "Version":"2012-10-17",
    "Statement":[  
        {  
            "Condition":{  
                "NumericLessThan":{  
                    "aws:MultiFactorAuthAge":"43200"
                }
            },
            "Action":"*",
            "Resource":"*",
            "Effect":"Allow"
        }
    ]
}
```

And the error:

```
aws iam create-access-key --profile test --user glenn.pratt

A client error (AccessDenied) occurred when calling the CreateAccessKey operation: User: arn:aws:iam::012345678901:user/glennpratt is not authorized to perform: iam:CreateAccessKey on resource: user glennpratt
```

If I add a role_arn, then aws-cli asks for an MFA token and assumes role, but we want to grant user permissions without assuming a role.

**Comment by jamesls** [comment]
Marking as a feature request.  The CLI currently does not support this, but it would be nice to have.

**Comment by ksperling** [comment]
Those waiting for this feature might want to have a look at https://github.com/lonelyplanet/aws-mfa (unfortunately it seems unmaintained).

One feature I would particularly like in this context is to be able to spawn a sub-shell via a command like `aws shell` such that the temporary credentials are only cached in the environment of that shell process rather than written to disk. This feature would also be beneficial in the context of `assume role`.

**Comment by ksperling** [comment]
I've implemented a shell script called [aws-session](https://github.com/ksperling/aws-session) that prompts for MFA credentials and makes them available in an interactive shell (think `sudo -s` for AWS).

At the moment it only supports STS GetSessionToken, but I may add AssumeRole support as well. One difference compared to the native AssumeRole support in the CLI is that the temporary credentials (intentionally) do not get persisted, so they can easily be discarded just by closing the shell.

**Comment by pwaller** [comment]
I've also written something called [`aws-creds`](https://github.com/pwaller/aws-creds) which caches credentials (or can be used to invoke a shell with the credentials in it) and also supports AssumeRole. I think there are a few of these out there, I already know two other organisations which have their own scripts for doing this. It would be great to see this implemented in the official tooling.

**Comment by ASayre** [comment]
Good Morning!

We're closing this issue here on GitHub, as part of our migration to [UserVoice](https://aws.uservoice.com/forums/598381-aws-command-line-interface) for feature requests involving the AWS CLI.

This will let us get the most important features to you, by making it easier to search for and show support for the features you care the most about, without diluting the conversation with bug reports.

As a quick UserVoice primer (if not already familiar): after an idea is posted, people can vote on the ideas, and the product team will be responding directly to the most popular suggestions.

We’ve imported existing feature requests from GitHub - Search for this issue there!

And don't worry, this issue will still exist on GitHub for posterity's sake.  As it’s a text-only import of the original post into UserVoice, we’ll still be keeping in mind the comments and discussion that already exist here on the GitHub issue.

GitHub will remain the channel for reporting bugs.  

Once again, this issue can now be found by searching for the title on: https://aws.uservoice.com/forums/598381-aws-command-line-interface 

-The AWS SDKs & Tools Team

**Comment by salmanwaheed** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  mkdirenv@gmail.com
    Domain salmanwaheed.info has exceeded the max emails per hour (186/150 (124%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from o8.sgmail.github.com ([167.89.101.199]:57580)
	by box1177.bluehost.com with esmtps (TLSv1.2:ECDHE-RSA-AES128-GCM-SHA256:128)
	(Exim 4.89_1)
	(envelope-from <bounces+848413-a7b0-hello=salmanwaheed.info@sgmail.github.com>)
	id 1ej0RO-001c37-GQ
	for hello@salmanwaheed.info; Tue, 06 Feb 2018 03:25:31 -0700
DKIM-Signature: v=1; a=rsa-sha1; c=relaxed/relaxed; d=github.com; 
	h=from:reply-to:to:cc:in-reply-to:references:subject:mime-version:content-type:content-transfer-encoding:list-id:list-archive:list-post:list-unsubscribe; 
	s=s20150108; bh=Fr89r4On8a1r6VY5SkJJNwy4Nx8=; b=toumNFHfeuc5+40X
	sUQLSduCNXTGyvYwL6pVxwTn5ZiEweoerDvNTBBec6XbIR2eutBHKSq+qFqrAcap
	2fQ30GzuVgK5cgSjJYeTLXb2SvvdgVxYxqFI/gGRc5Usv4jFk4MQiQsIgZJbxuw5
	D5sAFggmG2AzkhUlLX0SyxJptf8=
Received: by filter0035p1las1.sendgrid.net with SMTP id filter0035p1las1-16682-5A79828F-1A
        2018-02-06 10:25:19.672583041 +0000 UTC
Received: from github-smtp2a-ext-cp1-prd.iad.github.net (github-smtp2a-ext-cp1-prd.iad.github.net [192.30.253.16])
	by ismtpd0014p1iad2.sendgrid.net (SG) with ESMTP id ewEZaVmDRJ-aYMm9WnWRPA
	for <hello@salmanwaheed.info>; Tue, 06 Feb 2018 10:25:19.573 +0000 (UTC)
Date: Tue, 06 Feb 2018 10:25:19 +0000 (UTC)

**Comment by salmanwaheed** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  mkdirenv@gmail.com
    Domain salmanwaheed.info has exceeded the max emails per hour (187/150 (124%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
------ The body of the message is 6170 characters long; only the first
------ 5000 or so are included here.
Received: from github-smtp2-ext6.iad.github.net ([192.30.252.197]:60912 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by box1177.bluehost.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.89_1)
	(envelope-from <noreply@github.com>)
	id 1ej0RR-001c3U-9t
	for hello@salmanwaheed.info; Tue, 06 Feb 2018 03:25:33 -0700
Date: Tue, 06 Feb 2018 02:25:21 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1517912721;
	bh=F4Z/NdKuUHKikHp3+zaJPlQHGZiJ4WDvS4uXRxMSE9g=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=Q2NdxaajIIOP9c88o9Ln3Jwp0WrRFsVU2Ektsz4uHoHqNdxq+i5gMV2uKibxOX19L
	 mg86yb1gQk+DFaEaMDc3ipr8m3XEz/p0Vy3m9UxxscvfgQmMNPGrLzm7ZWvdoJfF3n
	 Uk8iy1e8O+Pn21zBMPiVfUSSDGY2w9SlQJdQMJiA=

**Comment by jamesls** [comment]
Based on community feedback, we have decided to return feature requests to GitHub issues.

**Comment by ahl** [comment]
Seems like a dupe of https://github.com/aws/aws-cli/pull/3174 and https://github.com/aws/aws-cli/pull/3174

**Comment by tb3088** [comment]
With AWS Professional Services incessantly pushing MFA and heavy use of account separation and assume-role and friends, the least the tools team could do is keep up, eh?

**Comment by tb3088** [comment]
this is a common IAM template fragment that the lack of support means things get annoying, fast. A more complete [example](https://gist.github.com/tb3088/83b3f8761456ace5ee3761cd613acd6b).

```
        {
            "Sid": "BlockMostAccessUnlessSignedInWithMFA",
            "Effect": "Deny",
            "NotAction": [
                "iam:GetAccountSummary",
                "iam:ListAccountAliases",
                "iam:ListUsers",
                "iam:ListServiceSpecificCredentials",
                "iam:ListMFADevices",
                "iam:EnableMFADevice",
                "iam:ResyncMFADevice",
                "iam:ListVirtualMFADevices",
                "iam:CreateVirtualMFADevice",
                "iam:DeleteVirtualMFADevice",
                "sts:GetSessionToken"
            ],
            "Resource": "*",
            "Condition": {
                "BoolIfExists": {
                    "aws:MultiFactorAuthPresent": "false"
                }
            }
        }
```

**Comment by cornfeedhobo** [comment]
Please implement this, like, yesterday.

**Comment by Angelin01** [comment]
6 years later and this is not supported. Thousands of mini shell scripts using `sts assume-role` and similar solutions just to get around this.

People will only implement extra protections if it is a good trade off between convenience and added security, and as of right now using MFA on the CLI is a huge inconvenience because of this and https://github.com/aws/aws-cli/pull/3174. Having worked with a lot of folks using AWS, I can count on my fingers how many permanently adopted MFA for the CLI, and of those who did everyone hacked together some sort of alternative solution to this exact issue.

**Comment by irl-segfault** [comment]
Agreed that it's strange to not have been fixed yet, however at least you *can* have MFA work (be prompted for) as long as the profile you're using assumes a role.  Temp credentials are stored in `~/.aws/cli/cache`, so you don't need a wrapper/scaffold script to update your `~/.aws/credentials` file. Having users interact with the API via a Role is a better practice anyways, though I feel everyone's pain.

**Comment by tb3088** [comment]
I do all that - look at cache, etc. in my scripts. The point is there should NOT be an assume-role required for MFA to be invoked. It's a bogus dependency that the API writers just jammed in there without thinking. The API writers have some glaring blind spots - namely naming consistency for starters.

**Comment by irl-segfault** [comment]
I don't disagree, just saying there's an easy workaround -- create a role, allow your user to assume the role, and add the roleARN to your `~/.aws/credentials` until this is fixed.  Agree it's not necessary / confusing.

**Comment by whereismypen** [comment]
Another quarter has lapsed and still nothing. No update, no progress, not even a promise. Even Google supports MFA on gcloud (with token even 👍 ).

**Comment by SebastianGogoasa** [comment]
It's becoming embarrassing 🤦.
A 6 years old feature request, oh dear...

**Comment by whereismypen** [comment]
Another quarter and this hasn't made any progress. Decision makers should consider this when choosing platforms. 
@SebastianGogoasa  - It's past embarrassing.

**Comment by SebastianGogoasa** [comment]
@whereismypen agree. I bet @glennpratt who originally raised this issue back in 2016 is very close to its retirement age.

I may be wrong but my guess is that AWS is working on a DNA based type of authentication for its CLI and I don't think we'll need this feature anymore... no more MFA, ADFS and all that non-sense.

That was a joke, but seriously this is a matter of security. What's the point in setting up MFA for the console when most of the developers are using the CLI on a daily basis.

**Comment by gitnik** [comment]
Let's try it again. More than 6 months later

**Comment by cornfeedhobo** [comment]
After more life experience, here is my $0.02:

Any company that has a big enough budget to get AWS to assign developer hours, is not using MFA directly in this way.

If you feel differently and work at a big company, then you will need to email this issue to your account's solution architect. That is the only way this will get attention. Don't hesitate. That's what they are there for.

**Comment by kjpgit** [comment]
Just remember at AWS, security is job 1.  As shown by this awesome ticket.

**Comment by ashovlin** [comment]
Hey all, in [2.30.3](https://github.com/aws/aws-cli/blob/v2/CHANGELOG.rst#2303), via https://github.com/aws/aws-cli/pull/9590, we released  `aws configure mfa-login` to create or update a profile with credentials acquired via MFA, without assuming a different role.

Docs: https://docs.aws.amazon.com/cli/latest/reference/configure/mfa-login.html

Example:
```
aws configure mfa-login --profile aws_mfa_account --update-profile mfa-temp --serial-number arn:aws:iam::012345678910:mfa/authenticator

MFA token code: 123456
Temporary credentials written to profile 'mfa-temp'
Credentials will expire at 2025-09-19 18:06:10 UTC
To use these credentials, specify --profile mfa-temp when running AWS CLI commands
```

It does require that you run `aws configure mfa-login` again once the temporary credentials in the targeted profile expire. Commands run with `--profile mfa-temp` will not automatically prompt you to enter your MFA code again like the `role_arn` workflow would. 

We can iterate on `aws configure mfa-login` for the MFA experience without assuming a role, please open new issues with feedback.

**Comment by tb3088** [comment]
it's a shame it was implemented as a dodgy unintuitive manner instead of the simple ~4 line diff generated eons ago and also didn't need an obtuse command .
